docs/design/design.md

```diff
@@ -0,0 +1,30 @@
+#Introduction
+
+The installer gathers all user configuration and consolidates it in the install-config.yaml file. It then uses this configuration to generate a slew of manifests. The Installer is responsible for bootstrapping the cluster. The bootstrap process starts several OpenShift operators while providing configuration to them via these manifests. More information about the Bootstrap process can be found [here](https://github.com/openshift/installer/blob/master/docs/dev/bootstrap_services.md). The role of the Installer comes to an end after this point. These operators in turn use these manifests to generate the resulting cluster. It then becomes the responsibilty of these Operators to manage changes within the cluster on day-2.
+
+While designing new features for the Installer, it is important to adhere to some design principles so that we achieve the following outcomes in no specific order of their importance:
+1. the design of different Installer features are consistent
+2. the Installer is able to provide the best user experince
+3. communicating the Installer's design goals and principles while colloborating with other teams
+
+# Design Documentation
+
+This is an attempt to document the OpenShift Installer's existing design principles and articulate design goals to guide new feature development and in some cases inform bug fixes. This is expected to be a living document which is continually updated when we gain clarity around some of these principles.
+
+## Design Goals
+
+### Install methods
+
+The OpenShift Installer currently supports two installation types, UPI (User Provisioned Infrastructure) and IPI (Installer Provisionined Infrastructure). IPI is an opinionated method of installing OpenShift. The IPI installation method tries to automate a large portion of the OpenShift Install with the least amount of intervention from the user. While it tries to take care of supporting the most common install scenarios, it cannot be expected to support every install variation that a user wants to use for their cloud.
+
+This is where UPI offers greater flexibility required for highly specialized installs. In this install method, the user is expected to provide several Infrastructure components needed for the install on their own. For this reason, documentation surrounding the UPI flow are especially important.
```

Comment:
Letting the user edit the manifests created by the installer is another way flexibility is managed and the install-config API is kept small. For example, when we added the workload partitioning feature, we did not add any flags to support it to install-config because it was only used by a small number of customers, at first. 

What heuristics does the team use to decide when to add a configuration option and when to rely on having the user edit manifests?